base/iterators.jl

```diff
@@ -1164,8 +1164,15 @@ julia> [(x,y) for x in 0:1 for y in 'a':'c']  # collects generators involving It
 """
 flatten(itr) = Flatten(itr)
 
-eltype(::Type{Flatten{I}}) where {I} = eltype(eltype(I))
 eltype(::Type{Flatten{Tuple{}}}) = eltype(Tuple{})
+eltype(::Type{Base.Iterators.Flatten{T}}) where T = tuple_eltype_join(T)
+
+tuple_eltype_join(::Type{Tuple{T}}) where T = eltype(T)
+function tuple_eltype_join(::Type{T}) where {T <: Tuple}
+    E = eltype(Base.tuple_type_head(T))
+    E === Any ? Any : typejoin(E, tuple_eltype_join(Base.tuple_type_tail(T)))
+end
```

Comment:
https://github.com/JuliaLang/julia/blob/1c449b3a0097bdf236648fe57412c74f35effb3c/base/iterators.jl#L397 also takes the splatting approach. If this works, maybe we could then just inline the whole thing, like this?
```suggestion
eltype(::Type{Base.Iterators.Flatten{T}}) where T = Base.typejoin(map(eltype, fieldtypes(T))...)
```